### ISSUE TITLE
New lint that suggests the use of mixed type operations like wrapping_sub_unsigned

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#13715]
### What it does

Suggests to replace code that contains an as cast, with specialized functions from the std library.

### Advantage

Makes the intent more explicit.

### Drawbacks

Slightly longer code.

### Example

```rust
#![warn(clippy::all)]
#![warn(clippy::pedantic)]
#![warn(clippy::nursery)]
#![warn(clippy::style)]
fn main() {
    let a = 1_i64;
    let b = 2_u64;
    let c = 2_u32;
    let x = a - (b as i64);
    let y = a.wrapping_sub_unsigned(b);
    let z = a.wrapping_sub_unsigned(u64::from(c));
    println!("{x:?} {y:?} {z:?}");
}
```

This code gives a clippy::cast_possible_wrap warning on the (b as i64) part, but I think it should suggest to replace code like in the computation of x, with the code that computes y. Similar suggestions could be used for other mixed-type operations.

**Comment by ojeda** [comment]
`a - (b as i64)` is not equivalent, since it may panic.

Should the example say `a.wrapping_sub(b as i64)` instead?

**Comment by leonardo-m** [comment]
The idea here is to suggest the programmer uses of mixed-typed math functions like `wrapping_sub_unsigned`, that are mixed, to replace code that contains a cast. But you're right, they don't do the same thing in presence of overflows. So perhaps we should close this issue.

**Comment by ojeda** [comment]
Yeah, what I meant is that perhaps there could still be a lint to suggest the change for `a.wrapping_sub(b as i64)` instead of the operators (assuming there is no lint for that yet).
